core/fetcher/fetcher_test.go

```diff
@@ -339,6 +341,140 @@ func TestFetchBuilderProposer(t *testing.T) {
+	}
+
+	// Construct sync committee selections.
+	selectionA := testutil.RandomSyncCommitteeSelection()
+	selectionA.SubcommitteeIndex = subCommIdxA
+	selectionB := testutil.RandomSyncCommitteeSelection()
+	selectionB.SubcommitteeIndex = subCommIdxB
+	commSelectionsByPubkey := map[core.PubKey]core.SignedData{
+		pubkeysByIdx[vIdxA]: core.NewSyncCommitteeSelection(selectionA),
+		pubkeysByIdx[vIdxB]: core.NewSyncCommitteeSelection(selectionB),
+	}
+
+	// Construct sync committee messages.
+	msgA := testutil.RandomSyncCommitteeMessage()
+	msgA.BeaconBlockRoot = beaconBlockRootA
+	msgB := testutil.RandomSyncCommitteeMessage()
+	msgB.BeaconBlockRoot = beaconBlockRootB
+	syncMsgsByPubkey := map[core.PubKey]core.SignedData{
+		pubkeysByIdx[vIdxA]: core.NewSignedSyncMessage(msgA),
+		pubkeysByIdx[vIdxB]: core.NewSignedSyncMessage(msgB),
+	}
+
+	// Construct beaconmock.
+	bmock, err := beaconmock.New(
+		beaconmock.WithSyncCommitteeSize(syncCommSize),
+		beaconmock.WithSyncCommitteeSubnetCount(syncCommSize),
+		beaconmock.WithTargetAggregatorsPerSyncSubcommittee(syncCommSize),
+	)
+	require.NoError(t, err)
+
+	bmock.SyncCommitteeContributionFunc = func(ctx context.Context, resSlot eth2p0.Slot, subcommitteeIndex uint64, beaconBlockRoot eth2p0.Root) (*altair.SyncCommitteeContribution, error) {
+		require.Equal(t, eth2p0.Slot(slot), resSlot)
+
+		var (
+			signedMsg       core.SignedSyncMessage
+			signedSelection core.SyncCommitteeSelection
+		)
+		for _, msg := range syncMsgsByPubkey {
+			m, ok := msg.(core.SignedSyncMessage)
+			require.True(t, ok)
+			if m.BeaconBlockRoot == beaconBlockRoot {
+				signedMsg = m
+			}
+		}
+		require.NotNil(t, signedMsg)
+
+		for _, selection := range commSelectionsByPubkey {
+			s, ok := selection.(core.SyncCommitteeSelection)
+			require.True(t, ok)
+			if s.SubcommitteeIndex == eth2p0.CommitteeIndex(subcommitteeIndex) {
+				signedSelection = s
+			}
+		}
+		require.NotNil(t, signedSelection)
+
+		return &altair.SyncCommitteeContribution{
+			Slot:              slot,
+			BeaconBlockRoot:   beaconBlockRoot,
+			SubcommitteeIndex: subcommitteeIndex,
+			AggregationBits:   bitfield.Bitvector128(testutil.RandomBitList()),
+			Signature:         testutil.RandomEth2Signature(),
+		}, nil
+	}
+
+	// Construct fetcher component.
+	fetch, err := fetcher.New(bmock, "")
+	require.NoError(t, err)
+
+	fetch.RegisterAggSigDB(func(ctx context.Context, duty core.Duty, key core.PubKey) (core.SignedData, error) {
+		if duty.Type == core.DutyPrepareSyncContribution {
+			require.Equal(t, core.NewPrepareSyncContributionDuty(slot), duty)
+			return commSelectionsByPubkey[key], nil
+		}
+
+		if duty.Type == core.DutySyncMessage {
+			require.Equal(t, core.NewSyncMessageDuty(slot), duty)
+			return syncMsgsByPubkey[key], nil
+		}
+
+		return nil, errors.New("unsupported duty")
+	})
+
+	fetch.Subscribe(func(ctx context.Context, resDuty core.Duty, resDataSet core.UnsignedDataSet) error {
+		require.Equal(t, duty, resDuty)
```

Comment:
how do you know this code was called?

Reply:
are you testing the "no aggregator" usecase?

Reply:
i'm testing both `aggregator` and `non-aggregator` in separate subtests!